pkg/proto/eth_math.go

```diff
@@ -0,0 +1,104 @@
+package proto
+
+import (
+	"fmt"
+	"math/big"
+)
+
+// Common big integers often used
+var (
+	big0   = big.NewInt(0)
+	big1   = big.NewInt(1)
+	big2   = big.NewInt(2)
+	big256 = big.NewInt(256)
+)
+
+// Various big integer limit values.
+var (
+	tt256   = new(big.Int).Exp(big2, big256, nil)
+	tt256m1 = new(big.Int).Sub(tt256, big1)
+)
+
+const (
+	// number of bits in a big.Word
+	bigWordBits = 32 << (uint64(^big.Word(0)) >> 63)
+	// number of bytes in a big.Word
+	bigWordBytes = bigWordBits / 8
+)
+
+// hexOrDecimal256 marshals big.Int as hex or decimal.
+type hexOrDecimal256 big.Int
+
+// newHexOrDecimal256 creates a new hexOrDecimal256
+func newHexOrDecimal256(x int64) *hexOrDecimal256 {
+	return (*hexOrDecimal256)(big.NewInt(x))
+}
+
+// parseEthereumBig256 parses s as a 256 bit integer in decimal or hexadecimal syntax.
+// Leading zeros are accepted. The empty string parses as zero.
+func parseEthereumBig256(s string) (*big.Int, bool) {
+	if s == "" {
+		return new(big.Int), true
+	}
+	var bigint *big.Int
+	var ok bool
+	if len(s) >= 2 && (s[:2] == "0x" || s[:2] == "0X") {
```

Comment:
the function you're looking for is `strings.TrimPrefix(s, "0x")`
instead of `(s[:2] == "0x" || s[:2] == "0X")` and `bigint, ok = new(big.Int).SetString(s[2:], 16)`

Reply:
The main point is detect and trim `0x` or `0X` prefix. In your version I can't detect and trim this prefixes in the short and convenient way. I have to implement both variants of `strings.TrimPrefix(s, "0x")` and `strings.TrimPrefix(s, "0X")`.